rfcs/0007-filtering.md

```diff
@@ -0,0 +1,211 @@
+
+    static {
+        eobFieldSelection.put("patient", wildCard);
+        eobFieldSelection.put("provider", wildCard);
+        eobFieldSelection.put("organization", wildCard);
+        eobFieldSelection.put("facility", wildCard);
+        eobFieldSelection.put("type", wildCard);
+        eobFieldSelection.put("resourceType", wildCard);
+        eobFieldSelection.put("diagnosis", wildCard);
+        eobFieldSelection.put("procedure", wildCard);
+        eobFieldSelection.put("item", buildItemDef());
+        eobFieldSelection.put("careTeam", wildCard);
+        eobFieldSelection.put("identifier", wildCard);
+    }
+
+    private static SubfieldSelection buildItemDef() {
+        Map<String, SubfieldSelection> itemMap = new HashMap<>();
+        itemMap.put("service", wildCard);
+        itemMap.put("quantity", wildCard);
+        itemMap.put("servicedPeriod", wildCard);
+        itemMap.put("location", wildCard);
+        itemMap.put("careTeamLinkId", wildCard);
+        itemMap.put("sequence", wildCard);
+        return new SubfieldSelection(itemMap);
+    }
+    
+    // Entry point
+    public static void filterEOBs(List<ExplanationOfBenefit> eobs) {
+        for (ExplanationOfBenefit eob : eobs) {
+            trimFields(eob, eobFieldSelection);
+        }
+    }
+
+    private static void trimFields(Object objToTrim, Map<String, SubfieldSelection> fieldMapping) {
+        Field[] fields = objToTrim.getClass().getDeclaredFields();
+        for (Field field : fields) {
+            // Do not return this field, null it out.
+            if (!fieldMapping.containsKey(field.getName())) {
+                PropertyUtils.setProperty(objToTrim, field.getName(), null);
+                continue;
+            }
+
+            // Return the entire field in the results
+            SubfieldSelection selection = fieldMapping.get(field.getName());
+            if (selection.getPickMe() == WhichFields.ALL) {
+                continue;
+            }
+
+            // Return a subset of the fields of the object referenced.
+            trimFieldsThrowException(objToTrim, field, selection);
+        }
+    }
+
+    private static void trimFieldsThrowException(Object objToTrim, Field field, SubfieldSelection selection) {
+        try {
+            trimFields(field.get(objToTrim), selection.getExplictFields());
+        } catch (IllegalAccessException iae) {
+            throw new RuntimeException(iae);
+        }
+    }
+``` 
+
+### Proposed Solution: Unresolved Questions
+[Proposed Solution: Unresolved Questions]: #proposed-solution-unresolved-questions
+
+How much will this save us?
+ 
+Proposed profiling is invoking BFD with at least 10,000 EOB Requests by Patient ID (ExplanationOfBenefit?patient=) and to measure
+the aggregate time of completion of all calls.  In other words, sum up the time taken by each call over all calls.
+
+This test will be baselined on current implementation and then ran again after the solution is deployed.
```

Comment:
Perhaps we should add the current baseline performance metrics within the RFC as well as the testing method so that we can reproduce results when a proposed code change is tested in lower environments. We'll need to use synthetic data for these tests / metrics.